File: cpp/behavioral/visitor.cpp
/*!
* \brief   Visitor. Represent an operation to be performed 
*       on the elements of an object structure. Visitor lets 
*       you define a new operation without changing the classes
*       of the elements on which it operates.
*
*  <ObjectStructure>      <Element>                <Visitor>
*     School  (create)->   Engineer       <==>        HR
*                             |                       | |
*                    SoftwareEngineer     IncomeVisitor VacationVisitor
*/

#include <iostream>
#include <string>

class Engineer;

// Visitor.
class HR {
public:
  virtual void Visit(Engineer *element) {};
};

// Element
class Engineer  {
public:
  virtual void ShowIncome() = 0;
  virtual void ShowVacation() = 0;
  // Allow a visitor to access it.
  virtual void Accept(HR *visitor) {};
};

// ConcreteElement
class SoftwareEngineer : public Engineer {
public:
  SoftwareEngineer(std::string name, int income, int vacation) {
    this->name_ = name;
    this->income_ = income;
    this->vacation_ = vacation;
  }

  void Accept(HR *visitor) {
    visitor->Visit(this);
  }

  // Show something or do something.
  void ShowIncome() {
    std::cout << name_ << "'s income: " << income_ << std::endl;
  }
  void ShowVacation() {
    std::cout << name_ << "'s vacation: " << vacation_ << std::endl;
  }

public:
  std::string name_;
  int income_;
  int vacation_;
};

// ConcreteVisitor.
class IncomeVisitor : public HR {
public:
  void Visit(Engineer *element) {
    element->ShowIncome();
  }
};

// ConcreteVisitor.
class VacationVisitor : public HR {
public:
  void Visit(Engineer *element) {
    element->ShowVacation();
  }
};

// ObjectStructure. (Element producer).
// It is typically contains multiple containers of different types 
// and interfaces, such as List, Set, Map, etc.
class School {
public:
  static Engineer *CreateSoftwareEngineer(std::string name,
    int income, int vacation) {
    Engineer *t = new SoftwareEngineer(name, income, vacation);
    return t;
  }
};

void main() {
  // Create two visitors  
  HR *v1 = new IncomeVisitor();
  HR *v2 = new VacationVisitor();

  // Create Engineers by School.
  Engineer *Tom = School::CreateSoftwareEngineer("Tom", 25000, 14);
  Engineer *Jimmy = School::CreateSoftwareEngineer("Jimmy", 36000, 6);

  Tom->Accept(v1);
  Tom->Accept(v2);

  Jimmy->Accept(v1);
  Jimmy->Accept(v2);

  delete v1;
  delete v2;
  delete Tom;
  delete Jimmy;
}

File: python/behavioral/visitor.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

"""
brief:
  Visitor. 
    Represent an operation to be performed 
  on the elements of an object structure. Visitor lets 
  you define a new operation without changing the classes
  of the elements on which it operates.

  <ObjectStructure>      <Element>              <Visitor>
    School  (create)->   Engineer       <==>      HR
                            |                     | |
                  SoftwareEngineer     IncomeVisitor VacationVisitor
"""

# ConcreteVisitor1.
class IncomeVisitor:
  def visit(self, engineer):
    engineer.show_income()
    
# ConcreteVisitor2.    
class VacationVisitor:
  def visit(self, engineer):
    engineer.show_vacation()

# ConcreteElement
class SoftwareEngineer: 
  def __init__(self, name, income, vacation):
    self.name = name
    self.income = income
    self.vacation = vacation
    
  def accept(self, visitor):
    visitor.visit(self)
    
  def show_income(self):
    print("%s's income: %d" % (self.name, self.income))
    
  def show_vacation(self):
    print("%s's vacation: %d" % (self.name, self.vacation))
 
# ObjectStructure. (Element producer).
# It is typically contains multiple containers of different types 
# and interfaces, such as List, Set, Map, etc. 
class School:
  def create_software_engineer(self, name, income, vacation):
    return SoftwareEngineer(name, income, vacation)
    
    
if __name__ == '__main__':

  # Create two visitors  
  v1 = IncomeVisitor()
  v2 = VacationVisitor()

  # Create Engineers by School.
  THU = School()
  Tom = THU.create_software_engineer("Tom", 25000, 14);
  Jimmy = THU.create_software_engineer("Jimmy", 36000, 6);

  Tom.accept(v1);
  Tom.accept(v2);

  Jimmy.accept(v1);
  Jimmy.accept(v2);
  
  